pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,333 @@
+	return nil // TODO
+}
+
+func (c *combinedStatusResolution) compareCombinedStatus(status *v1alpha1.CombinedStatus) bool {
+	return false //TODO
+}
+
+// evaluateWorkStatus evaluates the workstatus per relevant statuscollector.
+// The function returns true if any statuscollector data is updated.
+func (c *combinedStatusResolution) evaluateWorkStatus(celEvaluator *celEvaluator,
+	workStatusWECName string, workStatusContent *runtime.RawExtension) (bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.collectionDestinations.Has(workStatusWECName) {
+		return false, nil // workstatus is not relevant to this combinedstatus resolution
+	}
+
+	updated := false
+	for statusCollectorName, scData := range c.statusCollectorNameToData {
+		changed, err := evaluateWorkStatusAgainstStatusCollectorLocked(celEvaluator, workStatusWECName,
+			workStatusContent, scData)
+		if err != nil {
+			runtime2.HandleError(fmt.Errorf("failed to evaluate workstatus against statuscollector (%s): %w",
+				statusCollectorName, err)) // TODO: handle
+			return false, nil
+		}
+
+		updated = updated || changed
+	}
+
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. If the workstatus
+// does not match the filter, no other clause is evaluated.
+// The function returns true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
+// The function assumes that the caller holds a lock over the combinedstatus
+// resolution.
+func evaluateWorkStatusAgainstStatusCollectorLocked(celEvaluator *celEvaluator, workStatusWECName string,
+	workStatusContent *runtime.RawExtension, scData *statusCollectorData) (bool, error) {
+	wsData, exists := scData.workStatusRefToData[workStatusWECName]
+
+	// evaluate filter to determine if the workstatus is relevant
+	if scData.Filter != nil {
+		eval, err := celEvaluator.Evaluate(*scData.Filter, workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate filter expression: %w", err)
+		}
+
+		if eval.Type().TypeName() != "bool" {
+			return false, fmt.Errorf("filter expression must evaluate to a boolean")
+		}
+
+		if !eval.Value().(bool) { // workstatus is not relevant
+			if exists { // remove the workstatus data if it exists
+				delete(scData.workStatusRefToData, workStatusWECName)
+				return true, nil
+			}
+
+			return false, nil
+		}
+	}
+
+	if !exists {
+		wsData = &workStatusData{}
+		scData.workStatusRefToData[workStatusWECName] = wsData
+	}
+	updated := false
+
+	// evaluate select
+	selectEvals := make(map[string]ref.Val)
+	for _, selectNamedExp := range scData.Select {
+		eval, err := celEvaluator.Evaluate(selectNamedExp.Def, workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate select expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.selectEval[selectNamedExp.Name]
```

Comment:
This equality test looks wrong. `eval` and the RHS are of type `ref.Val`. Isn't the right way to test using the `Equal` method of `ref.Val`? Invoking `.Value()` on one side and not the other seems suspicious to me.